go/azidentity/azure_identity_credential_adapter.go

```diff
@@ -0,0 +1,67 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package azidentity
+
+import (
+	"github.com/Azure/azure-sdk-for-go/sdk/azidentity"
+	"github.com/Azure/go-autorest/autorest"
+	"github.com/Azure/go-autorest/autorest/adal"
+)
+
+//defaultScope defined the default Scopes
+const defaultScope = "https://management.azure.com/.default"
+
+//AzureIdentityCredentialAdapter is the adapter interface
+type AzureIdentityCredentialAdapter interface {
+	NewBearerAuthorizer() *autorest.BearerAuthorizer
+}
+
+//AzureIdentityCredentialAdapterBase is the base adapter
+type AzureIdentityCredentialAdapterBase struct {
+	Provider adal.OAuthTokenProvider
+}
+
+//NewAzureIdentityCredentialAdapterBase returns AzureIdentityCredentialAdapterBase
+//provider: a type that implements OAuthTokenProvider interface
+func NewAzureIdentityCredentialAdapterBase(provider adal.OAuthTokenProvider) *AzureIdentityCredentialAdapterBase {
+	return &AzureIdentityCredentialAdapterBase{
+		Provider: provider,
+	}
+}
+
+func (adapter *AzureIdentityCredentialAdapterBase) setDefaultScopes(scopes []string) []string {
+	if scopes == nil || len(scopes) == 0 {
+		scopes = []string{defaultScope}
+	}
+	return scopes
+}
+
+//NewBearerAuthorizer returns a new BearerAuthorizer
+func (adapter *AzureIdentityCredentialAdapterBase) NewBearerAuthorizer() *autorest.BearerAuthorizer {
+	bearerAuthorizer := autorest.NewBearerAuthorizer(adapter.Provider)
+	return bearerAuthorizer
+}
+
+//DefaultAzureCredentialAdapter is DefaultAzureCredential adapter
+type DefaultAzureCredentialAdapter struct {
+	*AzureIdentityCredentialAdapterBase
+}
+
+// NewDefaultAzureCredentialAdapter returns AzureDefaultAzureCredentialAdapter
+// scopes: The list of scopes for which the token will have access
+func NewDefaultAzureCredentialAdapter(scopes []string) (AzureIdentityCredentialAdapter, error) {
```

Comment:
Updated version based on the rewritten adapter.

```go
// DefaultManagementScope is the default credential scope for Azure Resource Management.
const DefaultManagementScope = "https://management.azure.com//.default"

// DefaultAzureCredentialOptions contains credential and authentication policy options.
type DefaultAzureCredentialOptions struct {
	// DefaultCredential contains configuration options passed to azidentity.NewDefaultAzureCredential().
	// Set this to nil to accept the underlying default behavior.
	DefaultCredential *azidentity.DefaultAzureCredentialOptions

	// AuthenticationPolicy contains configuration options passed to the underlying authentication policy.
	// Setting this to nil will use the DefaultManagementScope when acquiring a token.
	AuthenticationPolicy *azcore.AuthenticationPolicyOptions
}

// NewDefaultAzureCredentialAdapter returns an implementation of AzureIdentityCredentialAdapter.
func NewDefaultAzureCredentialAdapter(options *DefaultAzureCredentialOptions) (AzureIdentityCredentialAdapter, error) {
	if options == nil {
		options = &DefaultAzureCredentialOptions{
			AuthenticationPolicy: &azcore.AuthenticationPolicyOptions{
				Options: azcore.TokenRequestOptions{
					Scopes: []string{DefaultManagementScope},
				},
			},
		}
	}
	chain, err := azidentity.NewDefaultAzureCredential(options.DefaultCredential)
	if err != nil {
		return nil, err
	}
	return NewAzureIdentityCredentialAdapter(chain, *options.AuthenticationPolicy), nil
}
```

Reply:
Agree, I have changed it according this, but if using AuthenticationPolicy(), Error handling goes wrong with wrong azure credential. So I think using gettoken() is better.